@types/fs-extra: Namespace '"fs"' has no exported member 'BaseEncodingOptions'

- [x] I tried using the `@types/fs-extra` package and had problems.
- [x] I tried using the latest stable version of tsc. https://www.npmjs.com/package/typescript
- [x] I understand that `@types` package versions [do not follow SemVer](https://github.com/DefinitelyTyped/DefinitelyTyped#how-do-definitely-typed-package-versions-relate-to-versions-of-the-corresponding-library).
- [ ] I have a question that is inappropriate for [StackOverflow](https://stackoverflow.com/).  (Please ask any appropriate questions there).
- [x] [Mention](https://github.com/blog/821-mention-somebody-they-re-notified) the authors (see `Definitions by:` in `index.d.ts`) so they can respond.
  - Authors: @Andarist

Sorry if this is user error. I'm an absolute ts newbie here. Running ts on an electron project that is now pulling in @types/fs-extra I'm getting the following errors:

```
tsc
node_modules/@types/fs-extra/index.d.ts:232:38 - error TS2694: Namespace '"fs"' has no exported member 'BaseEncodingOptions'.

232         (path: PathLike, options: fs.BaseEncodingOptions | BufferEncoding | undefined | null): Promise<string>;
                                         ~~~~~~~~~~~~~~~~~~~

node_modules/@types/fs-extra/index.d.ts:233:38 - error TS2694: Namespace '"fs"' has no exported member 'BufferEncodingOption'.

233         (path: PathLike, options: fs.BufferEncodingOption): Promise<Buffer>;
                                         ~~~~~~~~~~~~~~~~~~~~

node_modules/@types/fs-extra/index.d.ts:234:38 - error TS2694: Namespace '"fs"' has no exported member 'BaseEncodingOptions'.

234         (path: PathLike, options: fs.BaseEncodingOptions | string | undefined | null): Promise<string | Buffer>;
                                         ~~~~~~~~~~~~~~~~~~~


Found 3 errors.
```

@types/fs-extra: 9.0.8
Node: 10.24.0
ts: 4.2.2

So.. I'm also getting this, but I can't even find `@types/fs-extra` **anywhere** in my project. I did also try explicitly adding `@types/fs-extra` to my dependencies as mentioned above, but unfortunately this made no difference for me. 🤔 

My currently active NodeJS version is `12.14.1`, installed via NVM, and my test project in question is a newly generated Ionic app, with `@angular/core@11.2.0` and `@capacitor/core@2.4.7`..

```
$ npm ls @types/fs-extra
ionic-angular-capacitor-test@0.0.1 /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test
└── (empty)
```

Meanwhile... **My** problems only comes in when I try run the Electron app (after building web files and syncing with added platforms, naturally):

```
$ npx cap open @capacitor-community/electron

> @capacitor-community/electron@1.3.4 capacitor:open /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test/node_modules/@capacitor-community/electron
> node dist/cap-scripts.js open

⠏ Opening Electron platformError: undefined
✖ Opening Electron platform:
(node:42357) UnhandledPromiseRejectionWarning:
> ionic-angular-capacitor-test@1.0.0 electron:start /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test/electron
> npm run build && electron ./


> ionic-angular-capacitor-test@1.0.0 build /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test/electron
> tsc

node_modules/@types/fs-extra/index.d.ts(232,38): error TS2694: Namespace '"fs"' has no exported member 'BaseEncodingOptions'.
node_modules/@types/fs-extra/index.d.ts(233,38): error TS2694: Namespace '"fs"' has no exported member 'BufferEncodingOption'.
node_modules/@types/fs-extra/index.d.ts(234,38): error TS2694: Namespace '"fs"' has no exported member 'BaseEncodingOptions'.
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! ionic-angular-capacitor-test@1.0.0 build: `tsc`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the ionic-angular-capacitor-test@1.0.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/agreeff/.npm/_logs/2021-03-05T13_41_36_064Z-debug.log
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! ionic-angular-capacitor-test@1.0.0 electron:start: `npm run build && electron ./`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the ionic-angular-capacitor-test@1.0.0 electron:start script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /Users/agreeff/.npm/_logs/2021-03-05T13_41_36_149Z-debug.log

(node:42357) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
(node:42357) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
```

I've been fighting with this for a few hours now with no luck.. Any bright ideas?

> So.. I'm also getting this, but I can't even find `@types/fs-extra` **anywhere** in my project. I did also try explicitly adding `@types/fs-extra` to my dependencies as mentioned above, but unfortunately this made no difference for me. 🤔
> 
> My currently active NodeJS version is `12.18.4`, installed via NVM, and my test project in question is a newly generated Ionic app, with `@angular/core@11.2.0` and `@capacitor/core@2.4.7`..
> 
> ```
> $ npm ls @types/fs-extra
> ionic-angular-capacitor-test@0.0.1 /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test
> └── (empty)
> ```
> 
> Meanwhile... **My** problem only comes in when I try run the Electron app (after building web files and syncing with added platforms, naturally):
> 
> ```
> $ npx cap open @capacitor-community/electron
> 
> > @capacitor-community/electron@1.3.4 capacitor:open /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test/node_modules/@capacitor-community/electron
> > node dist/cap-scripts.js open
> 
> ⠏ Opening Electron platformError: undefined
> ✖ Opening Electron platform:
> (node:42357) UnhandledPromiseRejectionWarning:
> > ionic-angular-capacitor-test@1.0.0 electron:start /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test/electron
> > npm run build && electron ./
> 
> 
> > ionic-angular-capacitor-test@1.0.0 build /Users/agreeff/workspace/sandbox/ionic-angular-capacitor-test/electron
> > tsc
> 
> node_modules/@types/fs-extra/index.d.ts(232,38): error TS2694: Namespace '"fs"' has no exported member 'BaseEncodingOptions'.
> node_modules/@types/fs-extra/index.d.ts(233,38): error TS2694: Namespace '"fs"' has no exported member 'BufferEncodingOption'.
> node_modules/@types/fs-extra/index.d.ts(234,38): error TS2694: Namespace '"fs"' has no exported member 'BaseEncodingOptions'.
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! ionic-angular-capacitor-test@1.0.0 build: `tsc`
> npm ERR! Exit status 2
> npm ERR!
> npm ERR! Failed at the ionic-angular-capacitor-test@1.0.0 build script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/agreeff/.npm/_logs/2021-03-05T13_41_36_064Z-debug.log
> npm ERR! code ELIFECYCLE
> npm ERR! errno 2
> npm ERR! ionic-angular-capacitor-test@1.0.0 electron:start: `npm run build && electron ./`
> npm ERR! Exit status 2
> npm ERR!
> npm ERR! Failed at the ionic-angular-capacitor-test@1.0.0 electron:start script.
> npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /Users/agreeff/.npm/_logs/2021-03-05T13_41_36_149Z-debug.log
> 
> (node:42357) UnhandledPromiseRejectionWarning: Unhandled promise rejection. This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). To terminate the node process on unhandled promise rejection, use the CLI flag `--unhandled-rejections=strict` (see https://nodejs.org/api/cli.html#cli_unhandled_rejections_mode). (rejection id: 1)
> (node:42357) [DEP0018] DeprecationWarning: Unhandled promise rejections are deprecated. In the future, promise rejections that are not handled will terminate the Node.js process with a non-zero exit code.
> ```
> 
> I've been fighting with this for a few hours now with no luck.. Any bright ideas?

you can just **cd** to electron folder, and run ```npm install --save @types/fs-extra@9.0.7```

--- a/types/fs-extra/fs-extra-tests.ts
+++ b/types/fs-extra/fs-extra-tests.ts
@@ -299,3 +299,9 @@ fs.realpath('src', (err, resolved) => {
 fs.realpath.native('src');
 // $ExpectType Promise<Buffer>
 fs.realpath.native('src', 'buffer');
+// $ExpectType Promise<string>
+fs.realpath.native('src', { encoding: 'utf-8' });
+// $ExpectType Promise<Buffer>
+fs.realpath.native('src', { encoding: 'buffer' });
+// $ExpectType Promise<string>
+fs.realpath.native('src', { encoding: null });

--- a/types/fs-extra/index.d.ts
+++ b/types/fs-extra/index.d.ts
@@ -229,9 +229,9 @@ export function realpath(path: PathLike, cache?: { [path: string]: string }): Pr
 /* tslint:disable:unified-signatures */
 export namespace realpath {
     const native: {
-        (path: PathLike, options: fs.BaseEncodingOptions | BufferEncoding | undefined | null): Promise<string>;
-        (path: PathLike, options: fs.BufferEncodingOption): Promise<Buffer>;
-        (path: PathLike, options: fs.BaseEncodingOptions | string | undefined | null): Promise<string | Buffer>;
+        (path: PathLike, options: { encoding: BufferEncoding | null } | BufferEncoding | undefined | null): Promise<string>;
+        (path: PathLike, options: { encoding: "buffer" } | "buffer"): Promise<Buffer>;
+        (path: PathLike, options: { encoding: BufferEncoding | null } | string | undefined | null): Promise<string | Buffer>;
         (path: PathLike): Promise<string>;
     } & typeof fs.realpath.native;
 }

